File: src/kernel/testdata/repo/oopl/javascript/demo/cards2/FrenchDeck.js
/**
 * @fileoverview
 */
goog.module('demo.cards2.FrenchDeck');
goog.module.declareLegacyNamespace();
const Card = goog.require('demo.cards2.Card');
const Deck = goog.require('demo.cards2.Deck');
const {MetaFrenchDeck} = goog.require('demo.cards2.FrenchDeckMeta');

/**
 * demo.cards2.FrenchDeck
 *   https://en.wikipedia.org/wiki/French_playing_cards
 * @public
 */
class FrenchDeck extends Deck {

  /**
   * initializer
   * @param {!boolean} [jokers=false]
   *   If true, the deck includes high and low joker.
   */
  constructor(jokers=false) {
    super();
    // User-provided code follows.
    var cards = this.cards();
    for (var suit = 1; suit < 5; ++suit) {
      for (var rank = 1; rank < 14; ++rank) {
        var card = new Card(rank, suit);
        cards.push(card);
      }
    }
    if (jokers) {
      var lowjoker = new Card(0, 0);
      cards.push(lowjoker);
      var highjoker = new Card(14, 0);
      cards.push(highjoker);
    }
  };

  /**
   * demo.cards2.FrenchDeck.asStr
   *   Provide a string representation of a given card.
   *   
   * @param {?demo.cards2.Card} card
   * @param {!boolean} [full=false]
   *   If true, result is '<rank> of <suit>'. If false,
   *   result is two letters.
   * @return {!string}
   *     Testing to see how things work.
   * @public
   * @override 
   * @suppress {reportUnknownTypes}
   */
  asStr(card, full=false) {
    var meta = MetaFrenchDeck;  /** TODO(wmh): Get this.meta() working */
    var suits = meta.Suits();
    var ranks = meta.Ranks();
    var result;
    if (full) {
      result = ranks[card.rank()] + ' of ' + suits[card.suit()];
    } else {
      result = ranks[card.rank()][0] + suits[card.suit()][0];
    }
    return result;
  };

  /**
   * demo.cards2.FrenchDeck.meta
   * @return {?metax.root.ObjectMetaRoot}
   * @public
   * @override 
   */
  meta() {
    return MetaFrenchDeck;
  };
}
exports = FrenchDeck;

File: src/kernel/testdata/repo/oopl/cpp/demo/cards2/FrenchDeck.cc
#include "demo/cards2/FrenchDeck.h"

namespace demo {
namespace cards2 {


FrenchDeck::FrenchDeck(bool jokers) {
  auto& cards = this->cardsRef();
  for (int suit = 1; suit < 5; ++suit) {
    for (int rank = 1; rank < 14; ++rank) {
      Card* card = new Card(rank, suit);
      cards.push_back(card);
    }
  }
  if (jokers) {
    Card* lowjoker = new Card(0, 0);
    cards.push_back(lowjoker);
    Card* highjoker = new Card(14, 0);
    cards.push_back(highjoker);
  }
}

std::string FrenchDeck::asStr(demo::cards2::Card* card, bool full) {
  auto* meta = dynamic_cast<FrenchDeckMeta*>(this->meta());
  auto& suits = meta->Suits();
  auto& ranks = meta->Ranks();
  std::string result;
  if (full) {
    result.append(ranks[card->rank()]);
    result.append(" of ");
    result.append(suits[card->suit()]);
  } else {
    result.push_back(ranks[card->rank()][0]);
    result.push_back(suits[card->suit()][0]);
  }
  return result;
}

metax::root::ObjectMetaRoot* FrenchDeck::meta() {
  return MetaFrenchDeck;
}


}  // cards2
}  // demo
